ir/instr.cpp

```diff
@@ -836,86 +853,139 @@ static StateValue fm_poison(State &s, expr a, const expr &ap,
                    ty, fmath, rm, bitwise, flags_in_only, 1);
 }
 
+static StateValue uf_float(const string &name,
+                           const vector<StateValue> &args,
+                           const expr &res,
+                           FastMathFlags fmath = FastMathFlags(),
+                           bool is_commutative = false) {
+  
+  vector<expr> arg_values;
+  arg_values.reserve(args.size());
+  for (auto &arg : args) {
+    arg_values.push_back(arg.value);
+  }
+
+  auto value = expr::mkUF(name, arg_values, res);
+  if (is_commutative) {
+    assert(args.size() == 2);
+    value = value & expr::mkUF(name, {arg_values[1], arg_values[0]}, res);
```

Comment:
I don't think this is a sound over-approximation. Do you need commutativity in your examples?
Would an axiom stating that `forall x, y . fadd(x, y) = fadd(y, x)` suffice?

Reply:
The encoding comes from "SMT-based Translation Validation for Machine Learning Compiler" by Seongwon Bang, Seunghyeon Nam, Inwhan Chun, Ho Young Jhoo, and Juneyoung Lee. The idea is to encode commutativity as

$\mathrm{op}(x, y) = \mathrm{op}'(x, y) \mspace{0.3em} \mathrm{and} \mspace{0.3em} \mathrm{op}'(y, x)$

where $\mathrm{and}$ is the bitwise and operator and $\mathrm{op}'$ is an uninterpreted function. There is a proof of correctness in the [supplementary material](https://zenodo.org/records/6615676) for the paper.

Reply:
I added a comment referencing the paper.

Reply:
The proof is very vague and doesn't show equisatisfiability clearly.
Why not use `x <= y ? f(x,y) : f(y, x)`?

Reply:
I would prefer to keep the current encoding, as I know that it works well on our test cases. While I could try other encodings, that might of course take some time. So here is my attempt at constructing a proof. It is of course inspired by the proof given in the supplementary material, but uses a simpler construction for $op'$ / $f'$.

Please correct me if I am wrong, but since we are already over approximating, we only need to prove that for all **commutative** functions $\mathrm{op}: \mathrm{BV}^n \times \mathrm{BV}^n \rightarrow \mathrm{BV}^n$ there is a function $\mathrm{op}': \mathrm{BV}^n \times \mathrm{BV}^n \rightarrow \mathrm{BV}^n$ such that

$$\mathrm{op}(x, y) = \mathrm{op}'(x, y) \mspace{0.3em} \mathrm{and} \mspace{0.3em} \mathrm{op}'(y, x)$$

(in the context of SMT solving this ensures that no model/counterexample can get lost.)

**Proof:** Let $\mathrm{BV}^n$ be the set of all bit vectors of size $n$ and let $\mathrm{op}: \mathrm{BV}^n \times \mathrm{BV}^n \rightarrow \mathrm{BV}^n$ be a commutative function.

Let $\mathrm{op}'(x, y) := \mathrm{op}(x, y)$. Then 

$$
\mathrm{op}'(x, y) \mspace{0.3em} \mathrm{and} \mspace{0.3em} \mathrm{op}'(y, x)
= \mathrm{op}(x, y) \mspace{0.3em} \mathrm{and} \mspace{0.3em} \mathrm{op}(y, x)
$$

By commutativity of $\mathrm{op}$

$$
= \mathrm{op}(x, y) \mspace{0.3em} \mathrm{and} \mspace{0.3em} \mathrm{op}(x, y)
$$

By $\forall x \in \mathrm{BV}^n: x \mspace{0.3em} \mathrm{and} \mspace{0.3em} x = x$

$$
= \mathrm{op}(x, y)
$$

Thus we have shown that $\mathrm{op}'$ fullfills the requirement.